File: cpp/questions/findMaximizedCapital.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = capital.size();
        std::vector<std::pair<int, int>> pj;
        pj.reserve(n);
        for (int p=0; p<n; ++p) {
            pj.emplace_back(capital[p], profits[p]);
        }
        std::sort(pj.begin(), pj.end());
        std::priority_queue<int, std::vector<int>, std::less<>> phq;
        int p = 0;
        while (k--) {
            while (p < n && w >= pj[p].first) {
                phq.push(pj[p].second);
                ++p;
            }
            if (phq.empty()) {
                break;
            } else {
                w += phq.top();
                phq.pop();
            }
        }
        return w;
    }
};

File: python/questions/findMaximizedCapital.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n = len(capital)
        pj = [(capital[i], profits[i]) for i in range(n)]
        pj.sort(key = lambda x: x[0])
        phq = []
        p = 0
        while k > 0:
            k -= 1
            while p<n and w>=pj[p][0]:
                heapq.heappush(phq, -pj[p][1])
                p += 1
            if not phq:
                break
            w -= heapq.heappop(phq)
        return w
